### PR TITLE
Generate System.ServiceModel.dll shim

### LINKED ISSUE #1848: Need to generate System.ServiceModel.dll shim
cc: @joshfree @weshaggard @ericstj 

UAP Metapackage's closure for netcore50 (uap10.0) includded a System.ServiceModel.dll shim that was restored from the compatibility package. We don't build that package anymore, and in order to be compatible with netcore50 we need to build this shim again for uap10.1's closure. This shim should be built in here, similar to how in corefx we build the shims for the rest of classic assemblies. Once we add that build, we should package it somewhere (perhaps the System.Private.ServiceModel package) so that we can flatten it out in our uap10.1 MetaPackage. 

This will need to be fixed before we ship the new UAP metapackage.

### PR DESCRIPTION
**Pull request by StephenBonikowsky**
* With the initial upgrade of WCF to buildtoolsv2 all the CoreFx contents under src/shims was ported to WCF but disabled.
* This PR enables shims.proj and modifies it to only build a System.ServiceModel.dll
* Cleanup everything that does not apply to WCF.
* Remove ApiCompat.proj and related files.
* Fixes #1848

### CODE REVIEW

**Comment by StephenBonikowsky** [comment]
@joperezr I hope I got this generally right... it would take some time to fully understand everything being done in shims.proj, I'm hoping most of it is boilerplate and will function just as well for System.ServiceModel as it does in CoreFx for the classic assemblies.

dir.props was used for both shims.proj and ApiCompat.proj, looking through the history in CoreFx I think there are still some pieces that can be removed, but would like your input.

**Comment by joperezr** [comment]
cc: @weshaggard

**Comment by joperezr** [comment]
Looks very good in general, my only question is where do we plan to package it?

**Comment by StephenBonikowsky** [comment]
You had mentioned System.Private.ServiceModel, should we do that as part of this same PR?

**Comment by weshaggard** [comment]
>You had mentioned System.Private.ServiceModel, should we do that as part of this same PR?

I suspect that makes the mode sense.

**Comment by joperezr** [comment]
yeah, that would be good for runtime, but I see that at least for corefx we also include this shims on the ref assets, @weshaggard would we need them there as well?

**Comment by joperezr** [comment]
>  should we do that as part of this same PR?

I don't see why not, that way you can actually close the issue that was opened, otherwise we would still have that pending in order to fully close the issue.

**Comment by weshaggard** [comment]
>yeah, that would be good for runtime, but I see that at least for corefx we also include this shims on the ref assets, @weshaggard would we need them there as well?

The ref is an interesting question. @zhenlan are you guys still planning on merging all the WCF contracts into one package? I know we talked about it in the past but I'm not sure if that is the plan or not. If it is the plan then this is an easy answer but if not then we will likely just need to pick one of the existing contract packages and put it into there. Probably the lowest level one.

**Comment by zhenlan** [comment]
> @zhenlan are you guys still planning on merging all the WCF contracts into one package? 

We still have the plan but it will be something in the future. We are not ready to do it in this release.

> if not then we will likely just need to pick one of the existing contract packages and put it into there. Probably the lowest level one.

In this case, sounds like `System.ServiceModel.Primitives` will be the right one.

**Comment by StephenBonikowsky** [comment]
After adding a project reference to shims.proj in the S.P.SM.pkgproj I could find...
`"C:\OSS\WCF\bin\runtime\netcoreapp-Windows_NT-Debug-x64\System.ServiceModel.dll"`
Can I build an actual nugget package locally to verify that System.ServiceModel.dll is present in the actual package?

**Comment by joperezr** [comment]
Build.cmd -allConfigurations

**Comment by weshaggard** [comment]
I would put the ref and impl shim in the same package so System.ServiceModel.Primitives package. I would also not do a project reference to the shims project. Instead I would add an explicit file reference to the output path in the pkgproj. Similar to https://github.com/dotnet/wcf/blob/master/Packaging.props#L57, but just ensure that the full stack build always builds the shim project before the pkgproj.

**Comment by StephenBonikowsky** [comment]
@weshaggard Looking to add the ref shim and I saw this...
"C:\OSS\WCF\src\System.ServiceModel.Primitives\ref\System.ServiceModel.Primitives.csproj"
This shim is for netfx right?

**Comment by weshaggard** [comment]
Yes that will pull in the System.ServiceModel.Primitives shim for netfx as well as the other configurations. It won't pull in the System.ServiceModel shim.

**Comment by StephenBonikowsky** [comment]
I amended the PR with an update to netfxreference.props, I was getting a lot of build errors, I know I don't need all of these references, I'm cleaning them up but I wanted to get a good build so I could get packages.

@joperezr @weshaggard Do I need a System.ServiceModel.dll in ref assets as well as in implementation assets? I have added it as an explicit file reference in the System.ServiceModel.Primitives.pkgproj, is this enough to get it put in the output dir?
```
    <!-- Shim for System.ServiceModel.dll -->
    <File Condition="Exists('$(ServiceModelFacade)')"
          Include="$(ServiceModelFacade)" >
    </File>
```

I don't see System.ServiceModel.dll anywhere in the System.ServiceModel.Primitives package that I built, where should I see it? Here is the current dir structure...
```
|   LICENSE.TXT
|   System.ServiceModel.Primitives.nuspec
|   THIRD-PARTY-NOTICES.TXT
|   version.txt
|   [Content_Types].xml
|
+---lib
|   +---MonoAndroid10
|   |       _._
|   +---MonoTouch10
|   |       _._
|   +---net45
|   |       _._
|   +---net46
|   |       System.ServiceModel.Primitives.dll
|   +---net461
|   |       System.ServiceModel.Primitives.dll
|   +---netcore50
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard1.3
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard2.0
|   |       System.ServiceModel.Primitives.dll
|   +---portable-net45%2Bwin8%2Bwp8
|   |       _._
|   +---uap10.0.15138
|   |       System.ServiceModel.Primitives.dll
|   +---win8
|   |       _._
|   +---wp8
|   |       _._
|   +---xamarinios10
|   |       _._
|   +---xamarinmac20
|   |       _._
|   +---xamarintvos10
|   |       _._
|   \---xamarinwatchos10
|           _._
+---package
|   \---services
|       \---metadata
|           \---core-properties
|                   413b1145a347488ebe9d92372cf9dec8.psmdcp
+---ref
|   +---MonoAndroid10
|   |       _._
|   +---MonoTouch10
|   |       _._
|   +---net45
|   |       _._
|   +---net46
|   |       System.ServiceModel.Primitives.dll
|   +---net461
|   |       System.ServiceModel.Primitives.dll
|   +---netcore50
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard1.0
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard1.1
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard1.3
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard2.0
|   |       System.ServiceModel.Primitives.dll
|   +---portable-net45%2Bwin8%2Bwp8
|   |       _._
|   +---uap10.0.15138
|   |       System.ServiceModel.Primitives.dll
|   +---win8
|   |       _._
|   +---wp8
|   |       _._
|   +---xamarinios10
|   |       _._
|   +---xamarinmac20
|   |       _._
|   +---xamarintvos10
|   |       _._
|   \---xamarinwatchos10
|           _._
\---_rels
        .rels
```

**Comment by joperezr** [comment]
you should be seeing it in `lib\uap10.0.15138\System.ServiceModel.dll` and `ref\uap10.0.15138\System.ServiceModel.dll`

**Comment by weshaggard** [comment]
With what you have I would expect it to be in the root. If you are directly using File you will need to at least manually specify the TargetPath metadata (see similar things like https://github.com/dotnet/buildtools/blob/master/src/Microsoft.DotNet.Build.Tasks.Packaging/src/PackageFiles/Packaging.targets#L314). We will need to figure out how many places we need this shim to support and we will need a File item for each location.

**Comment by StephenBonikowsky** [comment]
My latest update has System.ServiceModel.dll in both of the location specified by @joperezr , is it correct that they are both the same file, the shims project shouldn't be generating a ref version and an impl version of System.ServiceModel.dll?

I am getting some package validation errors, are they expected? If so how do I suppress them?
```
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error : System.ServiceModel.Primitives should only contain a single compile asset for UAP,Version=v10.0.15138/win10
-x86. [C:\OSS\WCF\src\System.ServiceModel.Primitives\pkg\System.ServiceModel.Primitives.pkgproj]
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error : System.ServiceModel.Primitives should support API version 4.2.1.0 on UAP,Version=v10.0.15138/win10-x86 but
C:\OSS\WCF\bin\obj\shims\uap\facades\System.ServiceModel.dll was found to support 4.0.0.0. [C:\OSS\WCF\src\System.ServiceModel.Primitives\pkg\System.ServiceMod
el.Primitives.pkgproj]
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error : System.ServiceModel.Primitives should support API version 4.2.1.0 on UAP,Version=v10.0.15138/win10-x86 but
C:\OSS\WCF\bin\obj\shims\uap\facades\System.ServiceModel.dll was found to support 4.0.0.0. [C:\OSS\WCF\src\System.ServiceModel.Primitives\pkg\System.ServiceMod
el.Primitives.pkgproj]
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error : System.ServiceModel.Primitives has mismatched compile (4.2.1.0) and runtime (4.0.0.0) versions on UAP,Versi
on=v10.0.15138/win10-x86. [C:\OSS\WCF\src\System.ServiceModel.Primitives\pkg\System.ServiceModel.Primitives.pkgproj]
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error MSB4018: The "ValidatePackage" task failed unexpectedly.\r [C:\OSS\WCF\src\System.ServiceModel.Primitives\pkg
\System.ServiceModel.Primitives.pkgproj]
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error MSB4018: System.NullReferenceException: Object reference not set to an instance of an object.\r [C:\OSS\WCF\s
rc\System.ServiceModel.Primitives\pkg\System.ServiceModel.Primitives.pkgproj]
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.Packaging.ValidatePackage.ValidateSupport()\r [C:\OSS\WCF\src\Sys
tem.ServiceModel.Primitives\pkg\System.ServiceModel.Primitives.pkgproj]
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error MSB4018:    at Microsoft.DotNet.Build.Tasks.Packaging.ValidatePackage.Execute()\r [C:\OSS\WCF\src\System.Serv
iceModel.Primitives\pkg\System.ServiceModel.Primitives.pkgproj]
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
\r [C:\OSS\WCF\src\System.ServiceModel.Primitives\pkg\System.ServiceModel.Primitives.pkgproj]
C:\OSS\WCF\Tools\Packaging.targets(1103,5): error MSB4018:    at Microsoft.Build.BackEnd.TaskBuilder.<ExecuteInstantiatedTask>d__26.MoveNext() [C:\OSS\WCF\src\
System.ServiceModel.Primitives\pkg\System.ServiceModel.Primitives.pkgproj]
Command execution failed with exit code 1.
```

**Comment by joperezr** [comment]
> is it correct that they are both the same file, the shims project shouldn't be generating a ref version and an impl version of System.ServiceModel.dll

That is correct the way it is. The only point of this file is to re-route types into the right assembly, and even though we are not generating one for impl, it will be re-routed appropiately. For example, Imagine type B existed in S.SM.Primitives.dll, the way this would work is:
For Refs:
```
System.ServiceModel.dll (typeForwards) -> S.SM.Primitives.dll (type lives there)
```

For Impls:
```
System.ServiceModel.dll (typeForwards) -> S.SM.Primitives.dll (typeforwards) -> System.Private.ServiceModel.dll (type lives there)
```

File: src/System.ServiceModel.Primitives/pkg/System.ServiceModel.Primitives.pkgproj [hunk 1]
```diff
@@ -28,6 +31,14 @@
       <!-- Supported inbox by all ns1.0-1.2 frameworks -->
       <Value>.NETStandard,Version=v1.0;.NETStandard,Version=v1.1;.NETStandard,Version=v1.2</Value>
     </ValidatePackageSuppression>
+    <!-- Impl shim for System.ServiceModel.dll -->
```
Comment (weshaggard) [inline src/System.ServiceModel.Primitives/pkg/System.ServiceModel.Primitives.pkgproj]: I assume we will also want this for .NET Core as well correct?

Comment (zhenlan) [inline src/System.ServiceModel.Primitives/pkg/System.ServiceModel.Primitives.pkgproj]: In this case, should we put the shim in netstandard2.0 folder under lib in the package?

Comment (joperezr) [inline src/System.ServiceModel.Primitives/pkg/System.ServiceModel.Primitives.pkgproj]: same as in UWP, so you will want to put it in both `lib\netstandard2.0` and `ref\netstandard2.0`

**Comment by StephenBonikowsky** [comment]
Updated, tree looks like this...
```
C:.
|   LICENSE.TXT
|   System.ServiceModel.Primitives.nuspec
|   THIRD-PARTY-NOTICES.TXT
|   version.txt
|   [Content_Types].xml
|
+---lib
|   +---MonoAndroid10
|   |       _._
|   +---MonoTouch10
|   |       _._
|   +---net45
|   |       _._
|   +---net46
|   |       System.ServiceModel.Primitives.dll
|   +---net461
|   |       System.ServiceModel.Primitives.dll
|   +---netcore50
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard1.3
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard2.0
|   |       System.ServiceModel.dll
|   |       System.ServiceModel.Primitives.dll
|   +---portable-net45%2Bwin8%2Bwp8
|   |       _._
|   +---uap10.0.15138
|   |       System.ServiceModel.dll
|   |       System.ServiceModel.Primitives.dll
|   +---win8
|   |       _._
|   +---wp8
|   |       _._
|   +---xamarinios10
|   |       _._
|   +---xamarinmac20
|   |       _._
|   +---xamarintvos10
|   |       _._
|   \---xamarinwatchos10
|           _._
+---package
|   \---services
|       \---metadata
|           \---core-properties
|                   44ffad2a81214786a15718e6a80134ef.psmdcp
+---ref
|   +---MonoAndroid10
|   |       _._
|   +---MonoTouch10
|   |       _._
|   +---net45
|   |       _._
|   +---net46
|   |       System.ServiceModel.Primitives.dll
|   +---net461
|   |       System.ServiceModel.Primitives.dll
|   +---netcore50
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard1.0
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard1.1
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard1.3
|   |       System.ServiceModel.Primitives.dll
|   +---netstandard2.0
|   |       System.ServiceModel.dll
|   |       System.ServiceModel.Primitives.dll
|   +---portable-net45%2Bwin8%2Bwp8
|   |       _._
|   +---uap10.0.15138
|   |       System.ServiceModel.dll
|   |       System.ServiceModel.Primitives.dll
|   +---win8
|   |       _._
|   +---wp8
|   |       _._
|   +---xamarinios10
|   |       _._
|   +---xamarinmac20
|   |       _._
|   +---xamarintvos10
|   |       _._
|   \---xamarinwatchos10
|           _._
\---_rels
        .rels
```
I still can't merge due to the package validation errors, suggestions on the correct way to suppress them?
Maybe @ericstj can recommend a solution?

**Comment by zhenlan** [comment]
@weshaggard @joperezr looks like both of you have signed off the PR. Thank you! Do you have any suggestions how we can avoid the validation errors so this PR can be merged?

**Comment by weshaggard** [comment]
I don't know if we have a way to suppress just this error @ericstj will have to help with that but he is out right now.

For the time being at least you can set SkipValidatePackage (https://github.com/dotnet/buildtools/blob/d626270ac6ed7451744ad1813efee5a40b786787/src/Microsoft.DotNet.Build.Tasks.Packaging/src/PackageFiles/Packaging.targets#L1131) to skip the package validation.

**Comment by joperezr** [comment]
I'll get this branch locally and try to figure out the problem and try to fix it, if not, we can skip validation as @weshaggard suggests.

**Comment by joperezr** [comment]
when pulling and building I noticed that the shims project doesn't get built in the build all configurations leg, which is wrong. This means that even after merging this, System.Servicemodel.Primitives package won't have the shim on it.

**Comment by StephenBonikowsky** [comment]
Looking into that, I think netfx configuration may not be getting triggered with allconfigurations.

**Comment by joperezr** [comment]
pushed in few changes that should fix all errors and allConfigurations leg.

**Comment by StephenBonikowsky** [comment]
I built it locally, no validation errors and I confirmed that System.ServiceModel.dll is in both the netstandard2.0 and uap10 folders of the S.ServiceModel.Primitives package.

Are we good to merge?

**Comment by joperezr** [comment]
yup. merge away 👍

**Comment by joperezr** [comment]
FYI, looks like we now have a release/uwp6.0 branch, and this change would have to go in there as well. Please do port this PR to that branch. cc: @danmosemsft @zamont

**Comment by weshaggard** [comment]
>I built it locally, no validation errors and I confirmed that System.ServiceModel.dll is in both the netstandard2.0 and uap10 folders of the S.ServiceModel.Primitives package.

Why do we need both the netstandard2.0 and uap10 folder assets? 

>when pulling and building I noticed that the shims project doesn't get built in the build all configurations leg, which is wrong. This means that even after merging this, System.Servicemodel.Primitives package won't have the shim on it.

Given that I think it is probably best to remove the exists condition so that we get build failures if this is missing instead of later finding out that it is broken.

**Comment by joperezr** [comment]
> Why do we need both the netstandard2.0 and uap10 folder assets?

We are doing this on all packages (even in corefx OOB packages) since there is still no mapping in NuGet between NS2.0 and uap10.0.RS3.0

**Comment by joperezr** [comment]
> Given that I think it is probably best to remove the exists condition so that we get build failures if this is missing instead of later finding out that it is broken.

That is actually a very good point, we should do this

**Comment by StephenBonikowsky** [comment]
> Given that I think it is probably best to remove the exists condition so that we get build failures if this is missing instead of later finding out that it is broken.

I'll remove the check.

**Pull request merged**